File: src/IncidentAtlas.Web/ClientApp/src/types/incident.ts
// ClientApp/src/types/incident.ts


//export type ApiErrorBody =
//    | Record<string, unknown>
//    | unknown[]
//    | string
//    | number
//    | boolean
//    | null;

//export class ApiError extends Error {
//    public readonly status: number;
//    public readonly body?: ApiErrorBody;

//    constructor(message: string, status: number, body?: ApiErrorBody) {
//        super(message);
//        this.name = "ApiError";
//        this.status = status;
//        this.body = body;
//    }
//}

export class IncidentSummaryDto {
    public readonly incidentId: string
    public readonly title: string
    public readonly severity: number
    public readonly status: number
    public readonly createdAtUtc: Date
    public readonly latestEventAtUtc: Date
    public readonly eventCount: number

    constructor(incidentId: string, title: string, severity: number,
        status: number, createdAtUtc: Date, latestEventAtUtc: Date, eventCount: number)
    {
        this.incidentId = incidentId
        this.title = title
        this.severity = severity
        this.status = status
        this.createdAtUtc = createdAtUtc
        this.latestEventAtUtc = latestEventAtUtc
        this.eventCount = eventCount
    }
}

export class IncidentDetailDto {
    public readonly incidentId: string
    public readonly title: string
    public readonly severity: number
    public readonly status: number
    public readonly createdAtUtc: Date
    public readonly events: Array<IncidentEventDto>

    constructor(incidentId: string, title: string, severity: number,
        status: number, createdAtUtc: Date, events: Array<IncidentEventDto>) {
        this.incidentId = incidentId
        this.title = title
        this.severity = severity
        this.status = status
        this.createdAtUtc = createdAtUtc
        this.events = events
    }
}

export class IncidentEventDto {
    public readonly incidentEventId: string
    public readonly sequence: number
    public readonly occurredAtUtc: Date
    public readonly type: number
    public readonly title: string
    public readonly details?: string
    public readonly createdBy?: string

    constructor(incidentEventId: string, sequence: number, occurredAtUtc: Date,
        type: number, title: string, details?: string, createdBy?: string)
    {
        this.incidentEventId = incidentEventId
        this.sequence = sequence
        this.occurredAtUtc = occurredAtUtc
        this.type = type
        this.title = title
        this.details = details
        this.createdBy = createdBy
    }
}

export class AppendIncidentEventRequest {
    public readonly type: number
    public readonly occurredAtUtc: Date
    public readonly title: string
    public readonly details?: string
    public readonly createdBy?: string

    constructor(type: number, occurredAtUtc: Date, title: string,
        details?: string, createdBy?: string)
    {
        this.type = type
        this.occurredAtUtc = occurredAtUtc
        this.title = title
        this.details = details
        this.createdBy = createdBy
    }
}

export const IncidentEventType = {
    IncidentDeclared: 0,
    NoteAdded: 1,
    StatusChanged: 2,
    Mitigated: 3,
    Resolved: 4,
    Reopened: 5
} as const;

export type IncidentEventType = typeof IncidentEventType[keyof typeof IncidentEventType];

export const IncidentSeverity = {
    Sev0: 0,
    Sev1: 1,
    Sev2: 2,
    Sev3: 3,
    Sev4: 4
} as const;

export type IncidentSeverity = typeof IncidentSeverity[keyof typeof IncidentSeverity];

export const IncidentStatus = {
    Active: 0,
    Mitigated: 1,
    Resolved: 2
} as const;

export type IncidentStatus = typeof IncidentStatus[keyof typeof IncidentStatus];

File: src/IncidentAtlas.Domain/Entities/Incident.cs


using IncidentAtlas.Domain.Enums;

namespace IncidentAtlas.Domain.Entities;

public class Incident
{
    public Guid IncidentId { get; private set; }

    public string Title { get; private set; } = String.Empty;
    public IncidentSeverity Severity { get; private set; }
    public IncidentStatus Status { get; private set; }
    public DateTimeOffset CreatedAtUtc { get; private set; }

    private readonly List<IncidentEvent> _events = new();
    public IReadOnlyList<IncidentEvent> Events => _events;
    public byte[] RowVersion { get; private set; } = Array.Empty<byte>();


    private Incident(
        Guid incidentId,
        string title,
        IncidentSeverity severity,
        DateTimeOffset createdAtUtc
    )
    {
        IncidentId = incidentId;
        Title = title;
        Severity = severity;
        Status = IncidentStatus.Active;
        CreatedAtUtc = createdAtUtc;
    }

    private Incident(
        Guid incidentId,
        string title,
        IncidentSeverity severity,
        IncidentStatus incidentStatus,
        DateTimeOffset createdAtUtc,
        IReadOnlyList<IncidentEvent> events
    )
    {
        IncidentId = incidentId;
        Title = title;
        Severity = severity;
        Status = incidentStatus;
        CreatedAtUtc = createdAtUtc;
        _events = events.ToList();
    }

    public static Incident Start(
        string title,
        IncidentSeverity severity,
        DateTimeOffset occurredAtUtc,
        string? details,
        string? createdBy
    )
    {
        if(string.IsNullOrWhiteSpace( title ))
        {
            throw new ArgumentException("Incident title is required.", nameof(title));
        }

        var incidentId = Guid.NewGuid();
        var eventId = Guid.NewGuid();

        var initialEvent = new IncidentEvent(
            incidentEventId: eventId,
            incidentId: incidentId,
            sequence: 1,
            occurredAtUtc: occurredAtUtc,
            type: IncidentEventType.IncidentDeclared,
            title: title,
            details: details,
            createdBy: createdBy,
            source: null
        );

        return new Incident(
            incidentId: incidentId,
            title: title,
            severity: severity,
            incidentStatus: IncidentStatus.Active,
            createdAtUtc: occurredAtUtc,
            events: new List<IncidentEvent> { initialEvent }
        );
    }

    public IncidentEvent AppendEvent(
        string title,
        IncidentEventType type,
        DateTimeOffset occurredAtUtc,
        string? details,
        string? createdBy
    )
    {
        var incidentEventId = Guid.NewGuid();

        var incidentEvent = new IncidentEvent(
            incidentEventId: incidentEventId,
            incidentId: IncidentId,
            sequence: Events.Count + 1,
            occurredAtUtc: occurredAtUtc,
            type: type,
            title: title,
            details: details,
            createdBy: createdBy,
            source: null
        );

        _events.Add(incidentEvent);

        switch(type)
        {
            case IncidentEventType.Mitigated: 
                Status = IncidentStatus.Mitigated;
                break;
            case IncidentEventType.Resolved:
                Status = IncidentStatus.Resolved;
                break;
            case IncidentEventType.Reopened:
                Status = IncidentStatus.Active;
                break;
        }

        return incidentEvent;
    }
}
